ERCS/erc-7786.md

```diff
@@ -0,0 +1,222 @@
+
+MAY generate and return a unique non-zero *outbox identifier*, otherwise returning zero. This identifier can be used to track the lifecycle of the message in the outbox in events and for post-processing.
+
+MUST emit a `MessageCreated` event, including the optional outbox identifier that is returned by the function.
+
+If an outbox identifier was generated, MUST emit a `MessageSent` event if no post-processing is required.
+
+#### `MessageCreated`
+
+This event signals that a would-be sender has initiated the sending of a message.
+
+If `outboxId` is present, post-processing MAY be required to send the message through the cross-chain channel.
+
+#### `MessageSent`
+
+This event signals that no more post-processing in the source chain is required, and that the message has been sent through the cross-chain channel.
+
+#### Post-processing
+
+After a sender has invoked `sendMessage`, further action MAY be required by the gateways to make the message effective. This is called *post-processing*. For example, some payment is typically required to cover the gas of executing the message at the destination.
+
+The exact interface for any such action is out of scope of this ERC. If the `postProcessingOwner` attribute is supported and present, such actions MUST be restricted to the specified account, otherwise they MUST be able to be performed by any party in a way that MUST NOT be able to compromise the eventual receipt of the message.
+
+The gateway MUST emit a `MessageSent` event with the appropriate identifier once no more post-processing is required by the source gateway. Further post-processing MAY be required in the source or destination gateways.
+
+### Destination Gateway
+
+A Destination Gateway is a contract that implements a protocol to validate messages sent on other chains and have them received at their destination.
+
+The gateway can operate in Active or Passive Mode.
+
+In both modes, the destination account of a message, aka the receiver, MUST implement a `receiveMessage` function. The use of this function depends on the mode of the gateway as described in the following sections.
+
+```solidity
+interface IERC7786Receiver {
+    function executeMessage(
+        address gateway,
+        bytes calldata gatewayMessageKey,
+        string calldata sourceChain, // CAIP-2 chain identifier
+        string calldata sender, // CAIP-10 account address
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) external payable;
+}
+```
+
+#### Active Mode
+
+The gateway directly invokes `receiveMessage`, and only does so with valid messages. The receiver MUST assume that a message is valid if the caller is a known gateway.
+
+The arguments `gateway` and `gatewayMessageKey` are unused in active mode and SHOULD be zero and empty respectively.
+
+#### Passive Mode
```

Comment:
I'm not a fan of the active / passive distinction, and don't really understand the passive mode's use case (cf. discussion on the comment thread in the motivation section above).

At first reading, I understood it as a way to distinguish between "permissioned" calls from a gateway operator, and "permissionless" calls for systems where proving can be done without permissionning things (light clients, zk). That doesn't seem to be the case, and even if it were, it could be achieve by letting people permissionlessly call a function on the gateway that then calls `receiveMessage`.

But the actual behaviour is ... not clear. The description makes it sound like it's validation only. This is a fair use case: a relayer may want to check if a message would succeed on the destination before relaying it. But can be achieved by either (1) an `eth_call`, even in permissioned cases (spoof the authorized sender, though this is a bit ugly/brittle and leaks implementation details), (2) a `verifyMessage` function (meant to be `eth_call`ed only) that verifies the non-permissioned parts (authenticity checks if any and app execution).

But maybe that's not what it does at all, as `setMessageExecuted` below seems to imply this is more than verification but actually stores an unexecuted message for later execution. This only seems useful insofar that the system is permissioned and the operator can eschew retrieving the source message from the source and only perform verification on it. But in a permissioned system, verification will require looking at the source anyway.

Reply:
I am ok with removing passive mode from the standard, to keep things simple